<a href="https://colab.research.google.com/github/cleysonl/Tensorflow2.0/blob/master/ANN.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ANN with TF 2.0

In [2]:
!pip install tensorflow-gpu==2.0.0.alpha0

     |████████████████████████████████| 332.1MB 84kB/s 
     |████████████████████████████████| 419kB 44.0MB/s 
     |████████████████████████████████| 3.0MB 31.8MB/s 


## Importing the libraries and the dataset

In [0]:
import numpy as np
import datetime
import tensorflow as tf
from tensorflow.keras.datasets import fashion_mnist

In [6]:
tf.__version__

'2.0.0-alpha0'

## Preprocessing the data

### Loading the Dataset

In [7]:
(X_train, y_train), (X_test, y_test) = fashion_mnist.load_data()

4423680/4422102 [==============================] - 0s 0us/step


### Normalizing the Dataset

We divide each pixel of the image in the training and test sets by the maximum number of pixels (255).

In [0]:
X_train = X_train / 255.0
X_test = X_test / 255.0

### Reshaping the dataset

In [0]:
# Since each image's dimension is 28x28, we reshape the full dataset to [-1 (all elements), height * width]
X_train = X_train.reshape(-1, 28*28)

In [0]:
X_train = X_train.reshape(-1,28*28)

In [11]:
# row = images and columns = pixels of the images
X_train.shape

(60000, 784)

In [0]:
X_test = X_test.reshape(-1,28*28)

In [14]:
# row = images and columns = pixels of the images
X_test.shape

(10000, 784)

## Building an ANN (Artificial Neural Network)

### Defining a Model

In [0]:
model = tf.keras.models.Sequential()

### Adding a first fully-connected hidden layer

In [0]:
model.add(tf.keras.layers.Dense(units = 128, activation = 'relu', input_shape=(784,)))

### Adding a second layer with dropout

In [0]:
model.add(tf.keras.layers.Dropout(0.2))

### Adding the output layer

In [0]:
model.add(tf.keras.layers.Dense(units=10,activation='softmax'))

### Compiling the model

Optimizer Adam

In [0]:
model.compile(optimizer='adam', loss='sparse_categorical_crossentropy', metrics = ['sparse_categorical_accuracy'])

In [21]:
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense (Dense)                (None, 128)               100480    
_________________________________________________________________
dropout (Dropout)            (None, 128)               0         
_________________________________________________________________
dense_1 (Dense)              (None, 10)                1290      
Total params: 101,770
Trainable params: 101,770
Non-trainable params: 0
_________________________________________________________________


### Training the model

In [22]:
model.fit(X_train, y_train, epochs=5)

Epoch 1/5
60000/60000 [==============================] - 5s 89us/sample - loss: 0.5297 - sparse_categorical_accuracy: 0.8121
Epoch 2/5
60000/60000 [==============================] - 5s 77us/sample - loss: 0.3991 - sparse_categorical_accuracy: 0.8550
Epoch 3/5
60000/60000 [==============================] - 5s 77us/sample - loss: 0.3672 - sparse_categorical_accuracy: 0.8672
Epoch 4/5
60000/60000 [==============================] - 5s 77us/sample - loss: 0.3462 - sparse_categorical_accuracy: 0.8740
Epoch 5/5
60000/60000 [==============================] - 5s 77us/sample - loss: 0.3305 - sparse_categorical_accuracy: 0.8777


### Model evaluation and prediction

In [23]:
test_loss, test_accuracy = model.evaluate(X_test, y_test)

10000/10000 [==============================] - 0s 44us/sample - loss: 0.3426 - sparse_categorical_accuracy: 0.8756


In [24]:
print("Test accuracy: {}".format(test_accuracy))

Test accuracy: 0.8755999803543091


# ***HOMEWORK: Artificial Neural Network***



*   Improve the performance of the current architecture and the goal is to reach 88% on the test set

